src/PotentialCalculation/SuccessiveOverRelaxation/SOR.jl

```diff
@@ -71,33 +72,55 @@ Thus, there cannot be local extrema.
 """
 @inline @fastmath function handle_depletion(
     new_potential::T, 
-    point_type::PointType, 
+    imp_scale::T,
     neighbor_potentials::NTuple{6,T}, 
     q_eff_imp::T, 
     volume_weight::T,
     sor_const::T
-)::Tuple{T, PointType} where {T}
+)::Tuple{T, T} where {T}
     vmin::T = min(neighbor_potentials...)
     vmax::T = max(neighbor_potentials...)
-
-    if new_potential < vmin || new_potential > vmax
-        new_potential -= q_eff_imp * volume_weight * sor_const
-        if (point_type & undepleted_bit == 0 && point_type & pn_junction_bit > 0) 
-            point_type += undepleted_bit
+    
+    is_ptype = q_eff_imp < 0 
+    neighbor_relevant_extremum = is_ptype ? vmin : vmax
+    imp_contribution = q_eff_imp * volume_weight * sor_const 
+    
+    if !iszero(imp_contribution) 
+        new_potential_proposal = new_potential - (T(1)- imp_scale) * imp_contribution
+        if (new_potential_proposal > vmax || new_potential_proposal < vmin)
+            imp_scale::T = (imp_contribution - (new_potential - neighbor_relevant_extremum)) / (imp_contribution * sor_const)
+            if imp_scale < 0 imp_scale = T(0) end
+            if imp_scale > 1 imp_scale = T(1) end
+            new_potential -= (T(1) - imp_scale) * imp_contribution 
+        else
+            new_potential = new_potential_proposal
         end
-    elseif point_type & undepleted_bit > 0
-        point_type -= undepleted_bit
     end
-    new_potential, point_type
+
+    new_potential, imp_scale
 end
 
 include("CPU_outerloop.jl")
 include("CPU_middleloop.jl")
 include("CPU_innerloop.jl")
-include("CPU_convergence.jl")
+include("CPU_update.jl")
 
 include("GPU_kernel_Cylindrical.jl")
 include("GPU_kernel_Cartesian3D.jl")
 include("GPU_kernel.jl")
-include("GPU_convergence.jl")
+include("GPU_update.jl")
+
+include("convergence.jl")
 
+
+function mark_undep_bits!(point_types::Array{PointType, 3}, imp_scale::Array{T, 3}) where {T}
+    @inbounds for i in eachindex(imp_scale)
+        if is_fixed_point_type(point_types[i])
+            imp_scale[i] = T(0)
+        end
+        if is_pn_junction_point_type(point_types[i]) && imp_scale[i] < 1
+            point_types[i] += undepleted_bit
+        end
```

Comment:
If we want to avoid this `if` statement:
```
point_types[i] += undepleted_bit * (is_pn_junction_point_type(point_types[i]) && imp_scale[i] < 1)
```

Reply:
I don't like this as it is hard to read.
But I tested it and is seems to be slower. (Though might depend on the simulation and grid.)